shub/image/deploy.py

```diff
@@ -79,51 +80,105 @@ def deploy_cmd(target, version, username, password, email,
-        while status in SYNC_DEPLOY_WAIT_STATUSES:
+class _BaseDeployProgress(object):
+
+    def __init__(self, status_url):
+        self.status_url = status_url
+
+    def show(self):
+        while True:
+            event = _check_status_url(self.status_url)
+            self.handle_event(event)
+            if event['status'] not in SYNC_DEPLOY_WAIT_STATUSES:
+                break
             time.sleep(SYNC_DEPLOY_REFRESH_TIMEOUT)
-            actual_state = _check_status_url(status_url)
-            if actual_state['status'] != status:
-                click.echo(" {}".format(actual_state))
-                status = actual_state['status']
-
-
-def _handle_deploy_errors(request):
-    content = request.json()
-    if request.status_code == 400 and content:
-        reason = content.get('non_field_errors')
-        if reason:
-            raise ShubException('\n'.join(reason))
-        else:
-            raise ShubException(request.content)
-    raise
+
+    def handle_event(self, event):
+        raise NotImplemented('Must be implemented in subclasses')
+
+
+class _LoggedDeployProgress(_BaseDeployProgress):
+    """Visualize deploy progress in verbose mode.
+
+    Output all the distinct events received from the service.
+    """
+    def __init__(self, status_url):
+        super(_LoggedDeployProgress, self).__init__(status_url)
+        self.previous_event = None
+        click.echo("Deploy results:")
+
+    def handle_event(self, event):
+        if event != self.previous_event:
+            click.echo("{}".format(event))
+            self.previous_event = event
+
+
+class _DeployProgress(_BaseDeployProgress):
+    """Visualize deploy progress in non-verbose mode.
+
+    Uses a progress bar to track total progress.
+    """
+    def __init__(self, status_url):
+        super(_DeployProgress, self).__init__(status_url)
+        self.progress_bar = self._create_progress_bar()
+        self.result_event = None
+
+    def show(self):
+        super(_DeployProgress, self).show()
+        # it's possible that release process finishes instantly without
+        # providing enough information to fill progress bar completely
+        if self.result_event and self.result_event['status'] == 'ok':
+            delta = max(self.progress_bar.total - self.progress_bar.n, 0)
+            self.progress_bar.update(delta)
+        self.progress_bar.close()
+        # last event with non-waiting status contains successful result or
+        # error result from the service with error details
+        if self.result_event:
+            click.echo("Deploy results:{}".format(self.result_event))
+
+    def handle_event(self, event):
+        if 'progress' in event and 'total' in event:
+            self.progress_bar.total = event['total']
+            self.progress_bar.update(max(event['progress'] - self.progress_bar.n, 0))
+        elif event['status'] not in SYNC_DEPLOY_WAIT_STATUSES:
+            self.result_event = event
```

Comment:
Good question, I'm thinking about cases when network is broken in the middle of the deploy process (or there's something wrong with releaser service itself) so last event will not reflect deploy results, I have doubts it's worth to print it anyway.